### PR TITLE
Feature/iFrame

### PR DESCRIPTION
**Pull request by sapieneptus**
**REBASED*** Tue Jan 12 19:11 CET

CC: @jmoody Integration tests for iFrame stuff. 

~~Currently failing because scrolling in WebViews is still broken, waiting on your fix.~~ FIX merged.

### CODE REVIEW

File: CalWebViewApp/CalWebViewApp/CalWebView-Info.plist [hunk 1]
```diff
@@ -2,6 +2,11 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>NSAppTransportSecurity</key>
+	<dict>
+		<key>NSAllowsArbitraryLoads</key>
+		<true/>
+	</dict>
```
Comment (jmoody) [inline CalWebViewApp/CalWebViewApp/CalWebView-Info.plist]: Can we not avoid this?

**Comment by jmoody** [comment]
I see the calabash.framework was updated.  I want to try to avoid updating the framework outside of a release, but I understand why you needed to here.  For a long time, I used to do arbitrary updates to the framework in CalSmokeApp repo.  Now that repo is HUGE which makes it a pain to clone.

We are also using this repo as a demonstration of the WebView API.  If we merge this before we release the iframe feature, users might be confused.

I set up the LPTestTarget to let us test server changes with cucumber.  I wonder if we should implement these tests there first?

I don't have a good solution.  Clearly, the final home of these changes is in this repo.

Any thoughts?

This is non-blocking.

File: CalWebViewApp/CalWebViewApp/CalWebView-Info.plist (hunk 1 above)
Comment (jmoody) [inline CalWebViewApp/CalWebViewApp/CalWebView-Info.plist]: This is non-blocking.

**Comment by jmoody** [comment]
**REBASED*** Tue Jan 12 19:11 CET

**Comment by jmoody** [comment]
### Tests

* [XTC](https://testcloud.xamarin.com/test/36e93161-eb29-45ec-aabb-4c6377276f13/) 34/44 failed

```shell
Failing Scenarios:
cucumber -p common features/iframe.feature:6 # Scenario: Query within a UIWebView iframe
cucumber -p common features/iframe.feature:13 # Scenario: Query within a WKWebView iframe
cucumber -p common features/iframe.feature:29 # Scenario: Query UIWebView iframe elements by id using css
cucumber -p common features/iframe.feature:38 # Scenario: Enter text in input fields within UIWebView iframe using css
cucumber -p common features/iframe.feature:45 # Scenario: Enter text in input fields within WKWebView iframe using css
cucumber -p common features/iframe.feature:52 # Scenario: Enter text in textarea within UIWebView iframe using css
cucumber -p common features/iframe.feature:57 # Scenario: Enter text in textarea within WKWebView iframe using css
cucumber -p common features/iframe.feature:68 # Scenario: Click a button within a WKWebView iframe
```

**Comment by sapieneptus** [comment]
Waiting to hear back on [this](https://github.com/calabash/calabash-ios-server/pull/308#issuecomment-171447026). It appears the javascript for UIWebViews doesn't work on WKWebViews. Unclear whether the answer is different javascript or that WKWebViews are more protective of x-domain iframes.

**Comment by jmoody** [comment]
I am resolving the merge conflicts and failing cucumbers.

**Comment by sapieneptus** [comment]
The final word from Safari debugger console is:

```
Blocked a frame with origin "null" from accessing a frame with origin "null". Protocols, domains, and ports must match.
```

I tried hosting page.html and iframe.html in the same location (raw.githubusercontent), but they explicitly block their content from being served in iframes. 

In conclusion, there's a _chance_ this will work in WKWebViews when the origins are the same, but to test it we'll need page.html and iframe.html to both be hosted on an iframe-friendly domain somewhere. Might just plop it on my personal server for now just to verify.

**Comment by jmoody** [comment]
Do you think that setting NSAppTransportSecurity might be causing a problem?  

> same location (raw.githubusercontent)

Can we remove the exception if we are hosting both with https?

**Comment by sapieneptus** [comment]
@jmoody yes we could, but we _can not_ host them both on raw.githubusercontent. So unless you know of a hosted https provider, for now I'm just going to throw them on my ec2 so we can just verify that it works for WKWebView.

**Comment by jmoody** [comment]
> ec2

The word on the street is that the SSL used in AWS domains is not compatible with Apple's requirements.  I spoke with an iOS developer about this last week.

> https://cdn.rawgit.com

Can we host one page at that URL and the other directly from GitHub?

And are you worried about the items mention here: https://rawgit.com/faq#why ?  Is it possible that the headers would make difference to the success or failure of JavaScript calls?  I am stabbing in the dark here.

**Comment by sapieneptus** [comment]
@jmoody EC2 are just linux boxes - you can have whatever kind of SSL you want on it. I think you're thinking of S3. 

But I don't want to put SSL on my ec2, I am just looking for a fast way to get both `page.html` and `iframe.html` on the same origin where `iframe.html` can be loaded into an iframe. raw.githubusercontent does not work because its content can't be displayed within an iframe.

If we host the pages in different places, then WKWebView will certainly fail. What I am testing is whether it works with both pages in the same domain (and local domain doesn't seem to work either - it reads the origin as `null`).

**Comment by jmoody** [comment]
> where to host

I have no other alternatives.

**Comment by jmoody** [comment]
@sapieneptus 

* Updated the server to 0.18.0 release
* Tagged the iframe scenarios - some of these need restarts to be effective
* Tagged three WKWebView iframe Scenarios as @pending 

I think that we can't link arbitrary versions of the server to test in-progress features.  This is a demonstration project and it can't serve both purposes:

* always passing
* testing in-progress features that require server updates

There are several merge conflict with the server and the size of the repo is ballooning - each server commit is 32 MB.

We should come up with a new strategy for testing in-progress server features. I recommend using the LPTestTarget in the calabash-ios-server.  I welcome any suggestions.  And to be clear, I was the one who recommended that we use this repo for testing the iframe feature.  In hindsight, I think it was the wrong decision.

Since this is a demo project, I think the iframe features should have been written more explicitly.

```
# Too generic
When I click the "button" with id "button"

# Specifically highlights that this is a Step that involves iframes
When I touch the submit button in the iframe
```

This will allow users to quickly find examples in the Steps.

I will be responsible for merging this PR once 0.18.0 has been release.

Great work on the iframe support.  Epic!

**Comment by jmoody** [comment]
@sapieneptus 

Checkout this failure:

* [iPad has 5 input fields showing](https://testcloud.xamarin.com/test/calwebview-cal_a73bc5ed-811e-453f-94aa-42fbaa6e4d6a/?step=2_0_2)

Should the `css:iframe` have filtered the other input fields out?

```
Then(/^I can query within the iframe with css for (\d+) "([^"]*)"$/) do |num, nodeName|
  page = page(WebViewApp::TabBar).active_page
  res = query(page.query_str("css:'iframe' css:'#{nodeName}'"))
  # Distinguish between a and a w/ href
  inputs = res.select { |elm| elm['nodeName'].downcase == nodeName.downcase }
  expect(inputs.count).to be == num.to_i
end
```

**Comment by sapieneptus** [comment]
Urm, whooops. Yes, it should have filtered those out. Wasn't thinking about big screens

**Comment by jmoody** [comment]
Is the problem on the server?

**Comment by sapieneptus** [comment]
Crap, maybe. I'll check this out after the meeting

**Comment by jmoody** [comment]
@sapieneptus 

I think there is no problem with server - it is the "ghost" iframe.  The contents are not being rendered, but the html is there.

All green:

* [XTC 1](https://testcloud.xamarin.com/test/calwebview-cal_0b5988a8-2618-4f6d-9fe5-144c9319aed2/)
* [XTC 2](https://testcloud.xamarin.com/test/calwebview-cal_d4d30b7e-28d2-406b-9217-c376705ab5f4/)

**Comment by jmoody** [comment]
@jonstoneman 

Once this runs through CI, I will merge it.  It will contain the Calabash iOS 0.18.0 server.

re: **Update iOS test apps to 0.18.0** [#790](https://github.com/xamarin/Xamarin.UITest/issues/790)

**Comment by jmoody** [comment]
@sapieneptus 

We cannot use githusercontent to serve the iframe in page.html.

We have to use the github cdn.

**Comment by jmoody** [comment]
Travis is acting flaky.

**Comment by jmoody** [comment]
Woot!

@jonstoneman You can now grab the sources to update iOS Test apps.

**Comment by sapieneptus** [comment]
enough is too much - I'm going to hound IT for somewhere we can host files with SSL

**Comment by jmoody** [comment]
FYI/BTW I tried to host the files on the gh-pages for this repo, but all traffic to the calabash GitHub organization is redirected to calaba.sh.  

I mention this because GitHub Pages _already have https_ support.

If we could get that redirect fixed, we could just host the page.html and iframe.html in the pages for this site.

**Comment by sapieneptus** [comment]
I am not sure that will work, at least with rawusercontent the pages are served with headers that disallow rendering within frames.

**Comment by krukow** [comment]
Guys, I can put static files in S3 if you want?

**Comment by jmoody** [comment]
S3 might not work.

I spoke with an iOS Developer about this - he tried S3 in an app and it did not work.  Amazon's SSL certs are not compatible with Apple's requirements.

He missed stand up today, but I have a TODO to follow up with him.

**Comment by sapieneptus** [comment]
I had the files on my private ec2, but there's no SSL so iOS wants to block the requests by default (and I soon all non-https requests will be blocked in iOS).

S3 (_supposedly_) doesn't work, Github doesn't work, ec2 without ssl is problematic.

**Comment by sapieneptus** [comment]
Localhost also doesn't work because iOS reads it as a `null` origin and apparently treats two `null` origins as different origins.

**Comment by jmoody** [comment]
GitHub is working?

Check master.  Both iframes render.

**Comment by sapieneptus** [comment]
rly? What did you do? I had them both hosted via raw.githubusercontent.com and they refused to render (because of the response headers disallowing frame rendering)

**Comment by jmoody** [comment]
I used the cdn.rawgit URL.

**Comment by jmoody** [comment]
https://testcloud.xamarin.com/test/414f8733-8e16-459d-86b0-0b4ce54c954f/?step=2_4_3&device=apple_ipad_air-7.1.1

**Comment by sapieneptus** [comment]
[It doesn't seem to be entering text...](https://testcloud.xamarin.com/test/414f8733-8e16-459d-86b0-0b4ce54c954f/?step=2_2_4&device=apple_ipad_air-7.1.1)

Is there any sort of javascript swizzling voodoo that goes on in XTC that might cause this (with the set_text.js file) ?

**Comment by jonstoneman** [comment]
If you end up needing another option then free-tier Azure Web apps might be an option: https://azure.microsoft.com/en-gb/pricing/details/app-service/

They support github continuous deployment.

**Comment by jmoody** [comment]
I did not notice that. Good catch.

This discussion has generated at least three issues:

1. Write the iframe Steps so they describe exactly what API is doing.
2. iframe support for WKWebView
3. Host the html files with https
4. What is text not appearing?

Where should these issues be captured? 

My first thought was here in this repo, but that will compound the issues-spread-across-repos problem we have already.  On the other hand, none of these are a priority at the moment.

@krukow What do you think?

**Comment by krukow** [comment]
@jmoody keep it here

**Pull request merged**